java/src/main/native/src/RmmJni.cpp

```diff
@@ -266,55 +364,232 @@ void set_java_device_memory_resource(
+        }
+    } else if (rmm::Manager::useManagedMemory()) {
+        auto tmp =  new rmm::mr::managed_memory_resource();
+        Initialized_resource.reset(tmp);
+        auto wrapped = make_tracking_adaptor(tmp, RMM_ALLOC_SIZE_ALIGNMENT);
+        Tracking_memory_resource.reset(wrapped);
+    } else {
+        auto tmp =  new rmm::mr::cuda_memory_resource();
+        Initialized_resource.reset(tmp);
+        auto wrapped = make_tracking_adaptor(tmp, RMM_ALLOC_SIZE_ALIGNMENT);
+        Tracking_memory_resource.reset(wrapped);
+    }
+    auto resource = Tracking_memory_resource.get();
+    rmm::mr::set_default_resource(resource);
+
+    std::unique_ptr<logging_resource_adaptor> log_result;
+    switch (log_to) {
+      case 1: // File
+        {
+          cudf::jni::native_jstring path(env, jpath);
+          std::ofstream * out = new std::ofstream(path.get());
+          log_result.reset(new logging_resource_adaptor(resource, out, true));
+        }
+        break;
+      case 2: // stdout
+        log_result.reset(new logging_resource_adaptor(resource, &std::cout, false));
+        break;
+      case 3: // stderr
+        log_result.reset(new logging_resource_adaptor(resource, &std::cerr, false));
+        break;
     }
-    rmmOptions_t opts;
-    opts.allocation_mode = static_cast<rmmAllocationMode_t>(allocation_mode);
-    opts.enable_logging = enable_logging == JNI_TRUE;
-    opts.initial_pool_size = pool_size;
-    JNI_RMM_TRY(env, , rmmInitialize(&opts));
-  } CATCH_STD(env, )
-}
 
-JNIEXPORT jboolean JNICALL Java_ai_rapids_cudf_Rmm_isInitializedInternal(JNIEnv *env, jclass clazz) {
-  try {
-    return rmmIsInitialized(nullptr);
-  } CATCH_STD(env, false)
+    if (log_result) {
+      if (Logging_memory_resource) {
+        JNI_THROW_NEW(env, RMM_EXCEPTION_CLASS, "Internal Error logging is double enabled", )
+      }
+
+      Logging_memory_resource = std::move(log_result);
+      auto replaced_resource = rmm::mr::set_default_resource(Logging_memory_resource.get());
+      if (resource != replaced_resource) {
+        rmm::mr::set_default_resource(replaced_resource);
+        Logging_memory_resource.reset(nullptr);
+        JNI_THROW_NEW(env, RMM_EXCEPTION_CLASS,
+            "Concurrent modification detected while installing memory resource", );
+      }
+    }
+  } CATCH_STD(env, )
 }
 
 JNIEXPORT void JNICALL Java_ai_rapids_cudf_Rmm_shutdownInternal(JNIEnv *env, jclass clazz) {
   try {
     set_java_device_memory_resource(env, nullptr, nullptr, nullptr);
-    JNI_RMM_TRY(env, , rmmFinalize());
+    // Instead of trying to undo all of the adaptors that we added in reverse order
+    // we just reset the base adaptor so the others will not be called any more
+    // and then clean them up in really any order.  There shoudl be no interaction with
+    // RMM during this time anyways.
+    Initialized_resource.reset(new rmm::mr::cuda_memory_resource());
+    rmm::mr::set_default_resource(Initialized_resource.get());
```

Comment:
Is it worth checking that the current default resource is the one we expect to be tearing down and throw/warn if it isn't?  Thinking of cases where the application code calls into separate native bits that interact with RMM and potentially install their own handler.

Reply:
It does not fix any race it just moves it.  RMM Init is saying go stomp on top of what is there and replace it with what I want. How can I tell if what is there is I am supposed to be replacing?

Reply:
I am thinking about this scenario:
# cudf inits RMM, installs handlers while remembering the original default handler
# app calls other native code outside of cudf that installs its own (potentially delegating?) default resource
# cudf shuts down RMM, installing the original default handler

At this point we've stomped on the app's default resource which could be bad, especially if the app's handler is going to delegate to our (now destroyed) resources and some resource allocated with the app resource is freed afterwards.  I was hoping to turn that undefined behavior into an assertion that when we go to shutdown, the current default resource is the one we last installed.  The java handler already does this check when it is deregistered, so I guess we're already covered there.

I admit this is a weird case and requires the app to reach around cudf to access RMM directly which seems unlikely.  Given it's already covered for the case when the Java handler is installed (which will be the common case), maybe nothing to do here.

Reply:
I don't think this is a valid use case.  When someone asks us to initialize RMM or to shut it down through Java or otherwise that says to me that we are going to wipe out all of the resource_managers and put the default resource manager in a known state. That is what we do. RMM internally was doing the same thing, which is what I based this off of.

https://github.com/rapidsai/rmm/blob/63ebb53bf21a58b98b4596f7b49a46d1d821b05d/src/memory_manager.cpp#L135-L152

https://github.com/rapidsai/rmm/blob/63ebb53bf21a58b98b4596f7b49a46d1d821b05d/src/memory_manager.cpp#L158-L167

There is no thought for someone else may have poked at the stack.  If you are really scared that someone is going to do that, then yes we can try and guard against it, but I would argue that it is beyond the scope of this because we are on par with what existed previously.